### PR TITLE
Feature/add haapi support

### PR DESCRIPTION
**Pull request by suren-khatana**
- Added HAAPI support for major flows
- Upgraded dependencies

### CODE REVIEW

File: src/main/java/io/curity/identityserver/plugin/usernamepassword/templates/ForgotPasswordPostRepresentation.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    private static String mask(String recipientOfCommunication)
+    {
+        if (StringUtils.isBlank(recipientOfCommunication))
+        {
+            // to prevent attackers phishing for valid email accounts, pretend an email was sent to some random address
+            recipientOfCommunication = RandomStringUtils.random(12, true, true);
+        }
+        Set<Integer> groupsToReplaceSet = new HashSet<>(Arrays.asList(2, 5));
+        Matcher matcher = Pattern.compile("(.*?)([^@]{1,4})(@)(.*?)([^.]{1,4})(\\.)?(.*)?").matcher(recipientOfCommunication);
```
Comment (daniellindau) [inline src/main/java/io/curity/identityserver/plugin/usernamepassword/templates/ForgotPasswordPostRepresentation.java:74]: This is a bit scary 😅
Maybe we can get a way with some simpler string operations.
```def recipient = "john.doe@acme.com"
def maskedRecipient = recipient.substring(0,4) + "********@" + recipient.substring(recipient.indexOf("@")+1, recipient.indexOf("@") + 5) + "***************"```

**Review by daniellindau — APPROVED** [review]

File: src/main/java/io/curity/identityserver/plugin/usernamepassword/templates/ForgotPasswordPostRepresentation.java (hunk 1 above)
Comment (suren-khatana) [reply src/main/java/io/curity/identityserver/plugin/usernamepassword/templates/ForgotPasswordPostRepresentation.java:74]: @daniellindau I agree that the code should be simplified. However, the suggested snippet does not handle email prefixes of 4 characters or fewer correctly. For example, if the email address is `kim@curity.io`, the masked result would be `kim@********@curi***************`, which is not as expected.

I suggest masking half of the prefix and half of the domain . I'll quickly implement this if it's okay, but I'm open to other suggestions.

Comment (suren-khatana) [reply src/main/java/io/curity/identityserver/plugin/usernamepassword/templates/ForgotPasswordPostRepresentation.java:74]:
```
   private static String mask(String recipientOfCommunication)
    {
        int atIndex = recipientOfCommunication.indexOf("@");
        String prefix = recipientOfCommunication.substring(0, atIndex);
        String domain = recipientOfCommunication.substring(atIndex + 1);

        // Masking one-third of the prefix
        int prefixHalfLength = Math.max(1, prefix.length() / 3);
        String maskedPrefix = prefix.substring(0, prefixHalfLength) + "****";

        // Masking one-third of the domain
        int domainHalfLength = Math.max(1, domain.length() / 3);
        String maskedDomain = domain.substring(0, domainHalfLength) + "****";

        return maskedPrefix + "@" + maskedDomain;
    }
```

**Comment by suren-khatana** [comment]
All of the updates from this PR has been copied to https://github.com/curityio/username-password-authenticator/pull/15, hence closing this PR.

**Pull request closed without merging**